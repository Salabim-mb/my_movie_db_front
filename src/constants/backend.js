const backend_path = process.env.REACT_APP_BACKEND_PATH;

export const backend = {
    MOVIES: backend_path + "movies/",
    USERS: backend_path + "user/",
    PLAIN: backend_path,

};