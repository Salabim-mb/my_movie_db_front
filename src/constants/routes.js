import MovieList from "../components/MovieList";

const paths = {
    MAIN: "/",
    LOGIN: "/login",
    REGISTER: "/register",
    MOVIES: "/movies",
    NEW_MOVIE: "/movies/new"
};

export default [
    {
        path: paths.MOVIES,
        exact: true,
        component: MovieList
    }r
]

export {paths};