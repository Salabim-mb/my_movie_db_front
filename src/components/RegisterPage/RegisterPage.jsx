import React, {useContext, useState} from "react";
import {Alert, Button, Card, Col, Form, Row} from "react-bootstrap";
import {backend} from "../../constants/backend";
import {getHeaders} from "../../utils/CORSHeaders";
import "./RegisterPage.css";
import {IndexLinkContainer} from 'react-router-bootstrap';
import {paths} from "../../constants/routes";
import {Redirect} from "react-router";
import {UserContext} from "../../context/UserContext";

const registerUser = async (data) => {
    let url = `${backend.PLAIN}register/`;
    let headers = getHeaders();

    let res = await fetch(url, {
        method: "POST",
        headers,
        body: JSON.stringify({
            userName: data.name,
            userSurname: data.surname,
            userEmail: data.email,
            userPassword: data.password
        })
    });

    if (res.status === 200 || res.status === 201) {
        return await res.json();
    } else {
        throw res.status;
    }
};

const RegisterPage = () => {
    const [data, setData] = useState({
        name: "",
        surname: "",
        email: "",
        password: "",
        password_R: ""
    });
    const [disabled, setDisabled] = useState(false);
    const [error, setError] = useState(false);
    const [validated, setValidated] = useState(false);
    const [correct, setCorrect] = useState(false);
    const [redirect, setRedirect] = useState(false);

    const user = useContext(UserContext);

    const submitForm = async (e) => {
        e.preventDefault();
        setError(false);
        setValidated(true);
        if (e.currentTarget.checkValidity() === false || data.password !== data.password_R) {
            e.stopPropagation();
        } else {
            setDisabled(true);
            try {
                let {userToken, userName, userSurname, userEmail} = await registerUser(data);
                let userData = {
                    name: userName,
                    surname: userSurname,
                    email: userEmail
                };
                user.login(userToken, userData);
                setCorrect(true);
                setTimeout(() => setRedirect(true), 3000);
            } catch(e) {
                console.log(e);
                setError(true);
            } finally {
                setDisabled(false);
            }
        }
    }

    const handlePassword = (e) => {
        setData({...data, [e.target.name]: e.target.value});
    };

    return (
        <Card className="card__container">
            <Card.Header className="card__header">
                <Card.Title as="h3">Sign up</Card.Title>
            </Card.Header>
            <Card.Body>
                <Form noValidate validated={validated} onSubmit={submitForm}>
                    <Row>
                        <Col>
                            <Form.Group controlId="firstname">
                                <Form.Label>First name:</Form.Label>
                                <Form.Control
                                    type="text"
                                    required
                                    placeholder="John"
                                    value={data.name}
                                    onChange={e => setData({...data, name: e.target.value})}
                                />
                                <Form.Control.Feedback type="invalid">
                                    First name must be provided
                                </Form.Control.Feedback>
                            </Form.Group>
                        </Col>
                        <Col>
                            <Form.Group controlId="lastname">
                                <Form.Label>Last name:</Form.Label>
                                <Form.Control
                                    type="text"
                                    placeholder="Doe"
                                    required
                                    value={data.surname}
                                    onChange={e => setData({...data, surname: e.target.value})}
                                />
                                <Form.Control.Feedback type="invalid">
                                    Last name must be provided
                                </Form.Control.Feedback>
                            </Form.Group>
                        </Col>
                    </Row>
                    <Form.Group controlId="email">
                        <Form.Label>Email:</Form.Label>
                        <Form.Control
                            type="email"
                            required
                            placeholder="johndoe@example.com"
                            value={data.email}
                            onChange={e => setData({...data, email: e.target.value})}
                        />
                        <Form.Control.Feedback type="invalid">
                            Correct email must be provided
                        </Form.Control.Feedback>
                    </Form.Group>
                    <Form.Group controlId="password">
                        <Form.Label>Password:</Form.Label>
                        <Form.Control
                            type="password"
                            minLength={8}
                            required
                            isInvalid={data.password && data.password_R && data.password !== data.password_R}
                            placeholder="Your password"
                            name="password"
                            value={data.password}
                            onChange={handlePassword}
                        />
                        <Form.Control.Feedback type="invalid">
                            Password must be at least 8 characters long
                        </Form.Control.Feedback>
                    </Form.Group>
                    <Form.Group controlId="passwordR">
                        <Form.Label>Repeat password:</Form.Label>
                        <Form.Control
                            type="password"
                            required
                            minLength={8}
                            isInvalid={data.password_R && data.password && data.password !== data.password_R}
                            placeholder="Your password again"
                            name="password_R"
                            value={data.password_R}
                            onChange={handlePassword}
                        />
                        <Form.Control.Feedback type="invalid">
                            Passwords must be equal
                        </Form.Control.Feedback>
                    </Form.Group>
                    <div className="justify-content-end d-flex">
                        <Button
                            type="submit"
                            disabled={disabled}
                            variant="success"
                        >
                            {disabled ? "Loading..." : "Sign me up!"}
                        </Button>
                    </div>
                    <IndexLinkContainer to={paths.LOGIN}>
                        <Card.Link>Already have an account? Sign in!</Card.Link>
                    </IndexLinkContainer>
                    {error && <Alert variant="danger">Something went wrong while trying to register user</Alert>}
                    {correct && <Alert variant="success">Registered successfully. Redirecting...</Alert>}
                    {redirect && <Redirect to={paths.USER_PAGE}/>}
                </Form>
            </Card.Body>
        </Card>
    )
};

export default RegisterPage;