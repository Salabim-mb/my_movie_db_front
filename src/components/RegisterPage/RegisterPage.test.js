import React from "react";
import {MemoryRouter} from "react-router";
import RegisterPage from "./RegisterPage";
import {createMemoryHistory} from 'history';
import {render, fireEvent, waitForElement} from '@testing-library/react';
import {Router} from "react-router-dom";
import {paths} from "../../constants/routes";

const renderWithRouter = (
    ui,
    {
        route = "/register",
        history = createMemoryHistory({ initialEntries: [route] })
    } = {}
) => {
    return {
        ...render(
            <Router history={history}>{ui}</Router>
        ),
        history
    }
};

describe("RegisterPage", () => {
    let apiFail;
    let token;
    let data;

    beforeAll(() => {
        global.fetch = jest.fn().mockImplementation((input, init) => {
            return new Promise((resolve, reject) => {
                if (apiFail) {
                    resolve({status: 500})
                } else {
                    resolve({status: 201, json: () => Promise.resolve({token: token})});
                }
            });
        });
    });

    beforeEach(() => {
        apiFail = false;
        jest.clearAllMocks();
        token = "123";
        data = {
            name: "Jan",
            surname: "Kowalski",
            email: "j.kowal@wp.pl",
            password: "zaq1@WSX",
            repeat_password: "zaq1@WSX"
        }
    });

    it("should match snapshot", () => {
        const {container} = render(
            <MemoryRouter>
                <RegisterPage />
            </MemoryRouter>
        );

        expect(container).toMatchSnapshot();
    });

    it("should block from posting if passwords dont match", () => {
        const {getByText, getByPlaceholderText} = render(
            <MemoryRouter>
                <RegisterPage/>
            </MemoryRouter>
        );

        fireEvent.change(getByPlaceholderText("John"), {
            target: {value: data.name}
        });
        fireEvent.change(getByPlaceholderText("Doe"), {
            target: {value: data.surname}
        });
        fireEvent.change(getByPlaceholderText("johndoe@example.com"), {
            target: {value: data.email}
        });
        fireEvent.change(getByPlaceholderText("Your password"), {
            target: {value: data.password}
        });
        fireEvent.change(getByPlaceholderText("Your password again"), {
            target: {value: data.repeat_password + "1"}
        });

        fireEvent.click(getByText("Sign me up!"));

        expect(getByText("Passwords must be equal")).toBeInTheDocument();
        expect(fetch).toHaveBeenCalledTimes(0);
    });

    it("should register correctly", async () => {
        const {history, getByText, getByPlaceholderText} = renderWithRouter(
            <MemoryRouter>
                <RegisterPage/>
            </MemoryRouter>
        );

        fireEvent.change(getByPlaceholderText("John"), {
            target: {value: data.name}
        });
        fireEvent.change(getByPlaceholderText("Doe"), {
            target: {value: data.surname}
        });
        fireEvent.change(getByPlaceholderText("johndoe@example.com"), {
            target: {value: data.email}
        });
        fireEvent.change(getByPlaceholderText("Your password"), {
            target: {value: data.password}
        });
        fireEvent.change(getByPlaceholderText("Your password again"), {
            target: {value: data.repeat_password}
        });

        fireEvent.click(getByText("Sign me up!"));

        await waitForElement(() => getByText("Registered successfully", {exact: false}));

        expect(getByText("Registered successfully", {exact: false})).toBeInTheDocument();
    });

    it("should view alert on api fail", async () => {
        apiFail = true;
        const {getByText, getByPlaceholderText} = render(
            <MemoryRouter>
                <RegisterPage/>
            </MemoryRouter>
        );

        fireEvent.change(getByPlaceholderText("John"), {
            target: {value: data.name}
        });
        fireEvent.change(getByPlaceholderText("Doe"), {
            target: {value: data.surname}
        });
        fireEvent.change(getByPlaceholderText("johndoe@example.com"), {
            target: {value: data.email}
        });
        fireEvent.change(getByPlaceholderText("Your password"), {
            target: {value: data.password}
        });
        fireEvent.change(getByPlaceholderText("Your password again"), {
            target: {value: data.repeat_password}
        });

        fireEvent.click(getByText("Sign me up!"));

        await expect(fetch).toHaveBeenCalledTimes(1);
        await waitForElement(() => getByText("Something went wrong", {exact: false}));

        expect(getByText("Something went wrong", {exact: false})).toBeInTheDocument();
    });

    it("should redirect to login page", () => {
        const {history, getByText} = renderWithRouter(
            <RegisterPage />
        );

        fireEvent.click(getByText("Already have an account?", {exact: false}));

        expect(history.location.pathname).toBe(paths.LOGIN);
    });
});