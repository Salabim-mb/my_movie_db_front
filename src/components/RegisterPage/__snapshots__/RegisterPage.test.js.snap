// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`RegisterPage should match snapshot 1`] = `
<div>
  <div
    class="card__container card"
  >
    <div
      class="card__header card-header"
    >
      <h3
        class="card-title"
      >
        Sign up
      </h3>
    </div>
    <div
      class="card-body"
    >
      <form
        class=""
        novalidate=""
      >
        <div
          class="row"
        >
          <div
            class="col"
          >
            <div
              class="form-group"
            >
              <label
                class="form-label"
                for="firstname"
              >
                First name:
              </label>
              <input
                class="form-control"
                id="firstname"
                placeholder="John"
                required=""
                type="text"
                value=""
              />
              <div
                class="invalid-feedback"
              >
                First name must be provided
              </div>
            </div>
          </div>
          <div
            class="col"
          >
            <div
              class="form-group"
            >
              <label
                class="form-label"
                for="lastname"
              >
                Last name:
              </label>
              <input
                class="form-control"
                id="lastname"
                placeholder="Doe"
                required=""
                type="text"
                value=""
              />
              <div
                class="invalid-feedback"
              >
                Last name must be provided
              </div>
            </div>
          </div>
        </div>
        <div
          class="form-group"
        >
          <label
            class="form-label"
            for="email"
          >
            Email:
          </label>
          <input
            class="form-control"
            id="email"
            placeholder="johndoe@example.com"
            required=""
            type="email"
            value=""
          />
          <div
            class="invalid-feedback"
          >
            Correct email must be provided
          </div>
        </div>
        <div
          class="form-group"
        >
          <label
            class="form-label"
            for="password"
          >
            Password:
          </label>
          <input
            class="form-control"
            id="password"
            minlength="8"
            name="password"
            placeholder="Your password"
            required=""
            type="password"
            value=""
          />
          <div
            class="invalid-feedback"
          >
            Password must be at least 8 characters long
          </div>
        </div>
        <div
          class="form-group"
        >
          <label
            class="form-label"
            for="passwordR"
          >
            Repeat password:
          </label>
          <input
            class="form-control"
            id="passwordR"
            minlength="8"
            name="password_R"
            placeholder="Your password again"
            required=""
            type="password"
            value=""
          />
          <div
            class="invalid-feedback"
          >
            Passwords must be equal
          </div>
        </div>
        <div
          class="justify-content-end d-flex"
        >
          <button
            class="btn btn-success"
            type="submit"
          >
            Sign me up!
          </button>
        </div>
        <a
          class="card-link"
          href="/login"
        >
          Already have an account? Sign in!
        </a>
      </form>
    </div>
  </div>
</div>
`;
